### ISSUE TITLE
Internal: Code Cleanup

### ISSUE THREAD

**Issue by jtopjian** [terraform-providers/terraform-provider-openstack#456]
This issue will track some internal code cleanup we plan to do over the next while. All PRs related to code cleanup should link to this issue.

At the moment, we plan to:

1. Replace provider-specific validation code with core Terraform helpers (where applicable).

2. Rename functions to use the "expand/flatten" pattern seen in other providers ([example](https://github.com/terraform-providers/terraform-provider-aws/blob/f0359b9048d0bfa6f5a6268d8d9ce071bbe54477/aws/resource_aws_eks_cluster.go#L251-L288)).

3. Attempt to standardize on log and error message formats.

4. Possibly move non-CRUD functions to a shared resource file. For example, functions in the `openstack_dns_*` resources that are not named `*Create`, `*Read`, `*Update`, or `*Delete` will be moved to a `dns_v2_shared.go` or `shared_dns_v2.go` file.

This list will be updated with any other cleanup items.

These changes _should not_ alter the resource's behavior. If we come across refactoring that might alter the behavior, it will be done in a separate issue. If a change introduced a bug, we fix it and create a new release as soon as possible.

**Comment by jtopjian** [comment]
@ozerovandrei I've updated this issue with some of the methods I used in the first few PRs. These aren't written in stone, though, so let me know if you want to see something changed.

**Comment by ozerovandrei** [comment]
@jtopjian rules looks good for now.

I think we will extend them with some additional cases about performing resource updates in several steps.
#258 is a good example of such complex updates.

I checked current PRs and I want to ask - can we also add simple unit tests for all new helpers in the `compute_flavor_v2.go`, `compute_keypair_v2.go` and `compute_secgroup_v2.go`? I think they will be helpful.

**Comment by kayrus** [comment]
According to [this comment](https://github.com/terraform-providers/terraform-provider-openstack/pull/515/files/9ac77a391917ba7341f8470cde7f24b5f5b7452e#r241261199) we can use `validation.StringInSlice` for limited input entries. Is a new ticket required for this task?

**Comment by kayrus** [comment]
Another [cleanup case](https://github.com/terraform-providers/terraform-provider-openstack/pull/515/files/9ac77a391917ba7341f8470cde7f24b5f5b7452e#r241262712):

remove `d.SetId("")` at the end of `Delete` functions.

**Comment by jtopjian** [comment]
> According to this comment we can use validation.StringInSlice for limited input entries. Is a new ticket required for this task?

No, it's listed in the first point at the beginning of this issue:

"Replace provider-specific validation code with core Terraform helpers (where applicable). Do not add new validation at this time!"

However, please do submit changes which span resources. All PRs should work on individual resources.

**Comment by jtopjian** [comment]
Note to self: before closing this issue, do one final sweep of `CheckDeleted` in the `Read` and `Delete` functions.

**Comment by kayrus** [comment]
CheckDeleted also should be added into the LBaaSv2 resources:

- openstack_lb_loadbalancer_v2
- openstack_lb_listener_v2
- openstack_lb_pool_v2
- openstack_lb_member_v2
- openstack_lb_monitor_v2
- openstack_lb_l7policy_v2
- openstack_lb_l7rule_v2

**Comment by jtopjian** [comment]
@kayrus #592 only updated resources which were previously cleaned up. The rest of the resources are in queue and we will get to them.

**Comment by kayrus** [comment]
new manila resources are affected too:

- openstack_sharedfilesystem_securityservice_v2
- openstack_sharedfilesystem_share_access_v2
- openstack_sharedfilesystem_sharenetwork_v2
- openstack_sharedfilesystem_share_v2

**Comment by jtopjian** [comment]
@kayrus Thank you, but there's no need to list them. They are in queue for being cleaned up, so all changes listed at the beginning of this issue will take place then.

**Comment by kayrus** [comment]
@jtopjian that was a note for myself.

**Comment by jtopjian** [comment]
@kayrus OK no problem :) I didn't want to see you researching work that was already known.

**Comment by kayrus** [comment]
```* openstack_compute_instance_v2.this (destroy): 1 error(s) occurred:

* openstack_compute_instance_v2.this: Error deleting OpenStack server: Resource not found
```

In my case the server was removed manually, therefore it is a minor issue.

Do I need to send a [PR](https://gist.github.com/kayrus/bbe33b9487418c8e89663e8d7d009eeb) or will it be covered by yet another cleanup task?

**Comment by jtopjian** [comment]
I've intentionally been ignoring `openstack_compute_instance_v2` until later. `openstack_compute_instance_v2` is similar to the port resource in that it's fragile and unique. 

If you'd like, please submit a PR with just the addition of `CheckDeleted`. 

With regard to the `admin_pass`, that's a schema change that wouldn't have been caught in a cleanup.  It's a small enough change, so you can bundle it in the same PR.
